app/src/main/java/com/github/bgabriel998/softwaredevproject/GeonamesHandler.java

```diff
@@ -1,99 +1,235 @@
-        //List containing all POIs around userLocation
-        ArrayList<POI> results = new ArrayList<>();
-        //Search POI close to the given location, using a limit of results and a given range
-        results = poiProvider.getPOICloseTo(userLocation,poiMaxResult,poiMaxRange);
-        return results;
+            throw new IllegalArgumentException("UserPoint user location can't be null");
+        if(boundingBoxRangeKm <= 0.1)
+            throw new IllegalArgumentException("BoundingBoxRangeKm can't be null or negative (also not under 100m)");
+        if(queryMaxResults < 1)
+            throw new IllegalArgumentException("QueryMaxResult parameter can't be less than 1");
+        if(queryTimeout <= 1)
+            throw new IllegalArgumentException("QueryTimeout parameter can't be less than 1 sec");
+
+        this.userLocation = userLocation;
+        this.rangeInKm = boundingBoxRangeKm;
+        this.queryMaxResults = queryMaxResults;
+        this.queryTimeout = queryTimeout;
+        this.poiProvider = new OverpassAPIProvider();
+        this.POIs = new ArrayList<POI>();
+    }
+
+
+
+    /**
+     * onPreExecute method.
+     * Setup bounding box for the POI query
+     * Creates the url query
+     */
+    @Override
+    protected void onPreExecute() {
+        super.onPreExecute();
+        BoundingBox boundingBox = userLocation.computeBoundingBox(rangeInKm);
+        queryUrl = poiProvider.urlForTagSearchKml("natural=peak", boundingBox,queryMaxResults,queryTimeout);
     }
 
     /**
-     * filterPOI : filter out point of interest list. The result list contains only peaks
-     * @param rawPOIList list containing all points of interests around Geopoint
-     * @return filtered list of points of interests containing only peaks
+     * See onPostExecute comments
+     * @param o
      */
-    private ArrayList<POI> filterPOI(ArrayList<POI> rawPOIList){
-        if(rawPOIList == null) throw new IllegalStateException("raw POI list is undefined");
-        //Arraylist containing only mountains POI
-        ArrayList<POI> results = new ArrayList<>();
-        for(POI point : rawPOIList){
-            //filter the poi's to get only the mountains
-            if(point.mCategory.equals("mountain")){
-                results.add(point);
-            }
+    @Override
+    protected void onPostExecute(Object o) {
+        super.onPostExecute(o);
+        if(o != null) {
+            //Filter out POI where the name or altitude is null
+            POIs = ((ArrayList<POI>) o).stream().filter(point -> point.mType != null && point.mLocation.getAltitude() != 0).collect(Collectors.toCollection(ArrayList::new));
+            onResponseReceived(POIs);
         }
-        return results;
+        else onResponseReceived(null);
+    }
+
+    /**
+     * See doInBackground comments
+     * @param objects nothing
+     * @return list of POI
+     */
+    @Override
+    protected Object doInBackground(Object[] objects) {
```

Comment:
What do we need the parameter for?
Why the returned type is Object and not ArrayList<POI>?

Reply:
The return type is `Object` since the method is an `@Override` method. The base method signature is located in the `AsyncTask`  class (android).
The return type is casted into object so that the method is suitable for all kinds of uses 
The parameter `objects` is not used